library/java/io/envoyproxy/envoymobile/Envoy.java

```diff
@@ -9,22 +9,67 @@
 // Wrapper class that allows for easy calling of Envoy's JNI interface in native Java.
 public class Envoy {
 
-  public void load() { System.loadLibrary("envoy_jni"); }
+  // Internal reference to helper object used to load and initialize the native library.
+  // Volatile to ensure double-checked locking works correctly.
+  private static volatile EnvoyLoader loader = null;
 
-  public void run(Context context, String config) {
-    Thread thread = new Thread(new Runnable() {
+  // Dedicated thread for running this instance of Envoy.
+  private final Thread runner;
+
+  // Private helper class used by the load method to ensure the native library and its
+  // dependencies are loaded and initialized at most once.
+  private static class EnvoyLoader {
+
+    EnvoyLoader(Context context) {
+      System.loadLibrary("envoy_jni");
+      initialize((ConnectivityManager)context.getSystemService(Context.CONNECTIVITY_SERVICE));
+    }
+  }
+
+  // Load and initialize Envoy and its dependencies, but only once.
+  public static void load(Context context) {
```

Comment:
I feel like I am missing something but shouldn't just making this method `synchronized` achieve the same behavior?

Reply:
Because we support lazy loading, making this synchronized requires this call to acquire a lock every time an Envoy instance is created - but the behavior we execute in here only ever has to happen once. If there was something that needed to happen every time, with high contention, synchronizing the full method would be more appropriate.

Reply:
Another way to look at is that as soon as we have a fully-created instance of `EnvoyLoader` we know the native library has been loaded and no longer need to synchronize.